adapters/Unity/UnityAdapter/GADMAdapterUnity.m

```diff
@@ -123,68 +131,74 @@ - (NSString *)getPlacementID {
 
-- (void)unityAdsPlacementStateChanged:(NSString *)placementId
+- (void)unityAdsPlacementStateChanged:(NSString *)placementID
                              oldState:(UnityAdsPlacementState)oldState
                              newState:(UnityAdsPlacementState)newState {
-  // This callback is not forwarded to the adapter by the GADMAdapterUnitySingleton and the adapter
-  // should use the unityAdsReady: and unityAdsDidError: callbacks to forward Unity Ads SDK state to
-  // Google Mobile Ads SDK.
+  id<GADMAdNetworkConnector> strongNetworkConnector = _networkConnector;
+  if (strongNetworkConnector && [placementID isEqualToString:_placementID]) {
+    if (newState == kUnityAdsPlacementStateNoFill){
+      NSError *errorWithDescription = GADUnityErrorWithDescription(@"NO_FILL");
+        [strongNetworkConnector adapter:self didFailAd:errorWithDescription];
+    }
+  }
 }
 
 - (void)unityAdsDidFinish:(NSString *)placementID withFinishState:(UnityAdsFinishState)state {
   id<GADMAdNetworkConnector> strongNetworkConnector = _networkConnector;
-  if (strongNetworkConnector) {
-    [strongNetworkConnector adapterWillDismissInterstitial:self];
-    [strongNetworkConnector adapterDidDismissInterstitial:self];
+  if (strongNetworkConnector && [placementID isEqualToString:_placementID]) {
+    if (state == kUnityAdsFinishStateCompleted) {
+     
+          
+    } else if (state == kUnityAdsFinishStateError) {
+        [strongNetworkConnector adapterWillDismissInterstitial:self];
+        [strongNetworkConnector adapterDidDismissInterstitial:self];
+    
+    } else if (state == kUnityAdsFinishStateSkipped) {
+        [strongNetworkConnector adapterWillDismissInterstitial:self];
+        [strongNetworkConnector adapterDidDismissInterstitial:self];
+    
+    }
+    [UnityAds removeDelegate:self];
   }
 }
 
 - (void)unityAdsReady:(NSString *)placementID {
   id<GADMAdNetworkConnector> strongNetworkConnector = _networkConnector;
-  if (!_isLoading) {
-    return;
-  }
-
-  if (strongNetworkConnector) {
+  if (strongNetworkConnector && [placementID isEqualToString:_placementID]) {
     [strongNetworkConnector adapterDidReceiveInterstitial:self];
   }
-  _isLoading = NO;
 }
 
 - (void)unityAdsDidClick:(NSString *)placementID {
   id<GADMAdNetworkConnector> strongNetworkConnector = _networkConnector;
   // The Unity Ads SDK doesn't provide an event for leaving the application, so the adapter assumes
   // that a click event indicates the user is leaving the application for a browser or deeplink, and
   // notifies the Google Mobile Ads SDK accordingly.
-  if (strongNetworkConnector) {
+  if (strongNetworkConnector && [placementID isEqualToString:_placementID]) {
     [strongNetworkConnector adapterDidGetAdClick:self];
     [strongNetworkConnector adapterWillLeaveApplication:self];
   }
 }
 
 - (void)unityAdsDidError:(UnityAdsError)error withMessage:(NSString *)message {
+//  id<GADMAdNetworkConnector> strongNetworkConnector = _networkConnector;
```

Comment:
So I see the `NO_FILL` handled above. What happens when other errors occur? For instance what if we are already showing an AD?

Reply:
`unityAdsDidFinish` with error status should be called.